src/libraries/Common/src/System/TimeProvider.cs

```diff
@@ -239,33 +239,36 @@ private static (uint duration, uint periodTime) CheckAndGetValues(TimeSpan dueTi
                 long dueTm = (long)dueTime.TotalMilliseconds;
                 long periodTm = (long)periodTime.TotalMilliseconds;
 
+                // MaxAllowedTimeout = 0xffffffff
+                // We have MaxSupportedTimeout = 0xfffffffe but we need to allow TimeSpan = Timeout.InfiniteTimeSpan which
+                // is -1. So we need to allow -1 milliseconds which equal to 0xffffffff.
+                const long MaxAllowedTimeout = (long)(unchecked((uint)Timeout.Infinite));
+
 #if SYSTEM_PRIVATE_CORELIB
                 ArgumentOutOfRangeException.ThrowIfLessThan(dueTm, -1, nameof(dueTime));
-                ArgumentOutOfRangeException.ThrowIfGreaterThan(dueTm, Timer.MaxSupportedTimeout, nameof(dueTime));
+                ArgumentOutOfRangeException.ThrowIfGreaterThan(dueTm, MaxAllowedTimeout, nameof(dueTime));
 
                 ArgumentOutOfRangeException.ThrowIfLessThan(periodTm, -1, nameof(periodTime));
-                ArgumentOutOfRangeException.ThrowIfGreaterThan(periodTm, Timer.MaxSupportedTimeout, nameof(periodTime));
+                ArgumentOutOfRangeException.ThrowIfGreaterThan(periodTm, MaxAllowedTimeout, nameof(periodTime));
```

Comment:
I don't understand these changes. dueTm and periodTm are signed. -1 will be -1, not 0xffffffff.

Reply:
I think this is a case of two wrongs making a right.  The problem isn't in CheckAndGetValues.  It's in the call chain of CheckAndGetValues passing in values that haven't been handled appropriately.  You can see in CTS for example that in the CancelAfter method it's special-casing:
https://github.com/dotnet/runtime/blob/a0964f9e3793cb36cc01d66c14a61e89ada5e7da/src/libraries/System.Private.CoreLib/src/System/Threading/CancellationTokenSource.cs#L450
but in the ctor it's not:
https://github.com/dotnet/runtime/blob/a0964f9e3793cb36cc01d66c14a61e89ada5e7da/src/libraries/System.Private.CoreLib/src/System/Threading/CancellationTokenSource.cs#L216

I think instead the right fix is to stop dealing with uints in the call sites, e.g.
https://github.com/dotnet/runtime/blob/a0964f9e3793cb36cc01d66c14a61e89ada5e7da/src/libraries/System.Private.CoreLib/src/System/Threading/CancellationTokenSource.cs#L164
and instead pass down either ints or the original TimeSpan instances.

Reply:
It is a good idea. I applied the suggestion. Also, I have removed all duration/period validation inside the TimeProvider timer. This is because we use the system timers which already validate the duration and the period.